### PR TITLE
8342011: Conditionally compile ReservedHeapSpace compressed heap support

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to ReservedHeapSpace to conditionally include the
support for compressed oops only in 64bit builds.  This code is unused and
inapplicable in 32bit builds.  Further, it contains code that is problematic
in a 32bit build.  See the JBS issue for details.

Testing: mach5 tier1, GHA testing

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342011](https://bugs.openjdk.org/browse/JDK-8342011): Conditionally compile ReservedHeapSpace compressed heap support (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [d9b0a011](https://git.openjdk.org/jdk/pull/21484/files/d9b0a011cb997b2b14065edb831ed7aacedf8357)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21484/head:pull/21484` \
`$ git checkout pull/21484`

Update a local copy of the PR: \
`$ git checkout pull/21484` \
`$ git pull https://git.openjdk.org/jdk.git pull/21484/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21484`

View PR using the GUI difftool: \
`$ git pr show -t 21484`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21484.diff">https://git.openjdk.org/jdk/pull/21484.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21484#issuecomment-2409489060)

### CODE REVIEW

File: src/hotspot/share/memory/virtualspace.cpp [hunk 1]
```diff
@@ -636,13 +641,16 @@ ReservedHeapSpace::ReservedHeapSpace(size_t size, size_t alignment, size_t page_
   guarantee(is_aligned(size, alignment), "set by caller");
 
   if (UseCompressedOops) {
+    NOT_LP64(ShouldNotReachHere();)
+#ifdef _LP64
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/virtualspace.cpp:645]: `#ifndef ... else ...` would look marginally better IMO.

**Review by dholmes-ora — APPROVED** [review]
One nit but looks good.

Thanks

File: src/hotspot/share/memory/virtualspace.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/memory/virtualspace.cpp]: Yeah, you're right.  Fixed.

File: src/hotspot/share/memory/virtualspace.hpp [hunk 2]
```diff
@@ -133,6 +133,9 @@ ReservedSpace ReservedSpace::partition(size_t offset, size_t partition_size)
 // Class encapsulating behavior specific of memory space reserved for Java heap.
 class ReservedHeapSpace : public ReservedSpace {
  private:
+
+  // Compressed oop support is not relevant in 32bit builds.
+#ifdef _LP64
```
Comment (stefank) [inline src/hotspot/share/memory/virtualspace.hpp:138]:
```suggestion

  // Compressed oop support is not relevant in 32bit builds.
#ifdef _LP64

```
The `#endif // _LP64 below have blank lines before and after, so it seems more consistent to also have it here.`

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/share/memory/virtualspace.cpp [hunk 3]
```diff
@@ -636,13 +641,17 @@ ReservedHeapSpace::ReservedHeapSpace(size_t size, size_t alignment, size_t page_
   guarantee(is_aligned(size, alignment), "set by caller");
 
   if (UseCompressedOops) {
+#ifndef _LP64
+    NOT_LP64(ShouldNotReachHere();)
+#else
```
[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/memory/virtualspace.cpp]: Yeah, forgot to remove the NOT_LP64.

Flipping the order seemed like it made the 32bit case coverage easy to miss.  But maybe that's okay. 
And it's not really all that far apart.  So okay, I've made that change.

**Review by dholmes-ora — APPROVED** [review]
LGTM

Thanks

File: src/hotspot/share/memory/virtualspace.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stefank) [reply src/hotspot/share/memory/virtualspace.cpp]: Thanks.

File: src/hotspot/share/memory/virtualspace.hpp (hunk 2 above)
Comment (stefank) [reply src/hotspot/share/memory/virtualspace.hpp:138]: Could you indicate if you saw this comment?

Comment (kimbarrett) [reply src/hotspot/share/memory/virtualspace.hpp:138]: Oops, missed that.  Fixed.

**Review by stefank — APPROVED** [review]
Looks good. Thanks.

**Comment by kimbarrett** [comment]
Thanks for reviews @dholmes-ora and @stefank

**Pull request closed without merging**